core-codemods/src/main/java/io/codemodder/codemods/SemgrepOverlyPermissiveFilePermissionsCodemod.java

```diff
@@ -0,0 +1,199 @@
+import com.github.javaparser.ast.expr.StringLiteralExpr;
+import com.github.javaparser.ast.stmt.ExpressionStmt;
+import io.codemodder.*;
+import io.codemodder.providers.sarif.semgrep.ProvidedSemgrepScan;
+import java.util.Optional;
+import javax.inject.Inject;
+
+@Codemod(
+    id =
+        "semgrep:java/java.lang.security.audit.overly-permissive-file-permission.overly-permissive-file-permission",
+    reviewGuidance = ReviewGuidance.MERGE_AFTER_CURSORY_REVIEW)
+public final class SemgrepOverlyPermissiveFilePermissionsCodemod
+    extends CompositeJavaParserChanger {
+
+  @Inject
+  public SemgrepOverlyPermissiveFilePermissionsCodemod(
+      final PermissionAddCallChanger addCallChanger,
+      final FromStringChanger fromStringChanger,
+      final InlineFromStringChanger inlineFromStringChanger) {
+    super(addCallChanger, fromStringChanger, inlineFromStringChanger);
+  }
+
+  /**
+   * This handles the case where the permission is added inline to {@code setPosixFilePermissions()}
+   * like:
+   *
+   * <p>{@code Files.setPosixFilePermissions(startupScript,
+   * PosixFilePermissions.fromString("rwxrwxrwx"));}
+   */
+  private static class InlineFromStringChanger
+      extends SarifPluginJavaParserChanger<ExpressionStmt> {
+    @Inject
+    public InlineFromStringChanger(
+        @ProvidedSemgrepScan(
+                ruleId =
+                    "java.lang.security.audit.overly-permissive-file-permission.overly-permissive-file-permission")
+            final RuleSarif sarif) {
+      super(
+          sarif,
+          ExpressionStmt.class,
+          RegionNodeMatcher.MATCHES_START,
+          CodemodReporterStrategy.empty());
+    }
+
+    @Override
+    public boolean onResultFound(
+        final CodemodInvocationContext context,
+        final CompilationUnit cu,
+        final ExpressionStmt stmt,
+        final Result result) {
+      Optional<MethodCallExpr> setPosixFilePermissionsRef =
+          expect(stmt)
+              .toBeExpressionStatement()
+              .withMethodCallExpression()
+              .withName("setPosixFilePermissions")
+              .withArgumentsSize(2)
+              .result();
+      if (setPosixFilePermissionsRef.isPresent()) {
+        MethodCallExpr setPosixFilePermissions = setPosixFilePermissionsRef.get();
+        Optional<MethodCallExpr> fromStringRef =
+            expect(setPosixFilePermissions.getArgument(1))
+                .toBeMethodCallExpression()
+                .withName("fromString")
+                .withArgumentsSize(1)
+                .result();
+        if (fromStringRef.isPresent()) {
+          return fixFromString(fromStringRef.get());
+        }
+      }
+      return false;
+    }
+  }
+
+  /**
+   * This handles the case where permissions are created from UNIX permission strings like:
+   *
+   * <p>{@code var p = Permissions.fromString("rwxrwxrwx")}
+   */
+  private static class FromStringChanger extends SarifPluginJavaParserChanger<ExpressionStmt> {
```

Comment:
although this class could still only be subclassed in this file, marking this as `final` still helps static analysis tools provide better suggestions 
```suggestion
  private static final class FromStringChanger extends SarifPluginJavaParserChanger<ExpressionStmt> {
```